### PR TITLE
Feature : duplicate a form

### PR DESCRIPTION
**Pull request by tonyyb**
Add an action on appdesk to duplicate forms. 
The duplication keep fields layout.

### CODE REVIEW

**Review by shaoshiva — APPROVED** [review]
Nice :+1:

File: classes/controller/admin/form.ctrl.php [hunk 1]
```diff
@@ -490,4 +490,129 @@ protected function getInputFieldData($name = 'fieldData')
...
+            \Response::json(array(
+                'dispatchEvent' => array(
+                    'name' => 'Nos\Form\Model_Form',
+                    'action' => 'insert',
+                    'context' => $context,
+                ),
+                'notify' => __('Here you are! The form has just been duplicated.'),
+            ));
+        } catch (\Exception $e) {
+            $this->send_error($e);
```
Comment (Meroje) [inline classes/controller/admin/form.ctrl.php]: isn't that already done automatically ?

File: classes/controller/admin/form.ctrl.php [hunk 2]
```diff
@@ -490,4 +490,129 @@ protected function getInputFieldData($name = 'fieldData')
...
+        $clone = clone $form;
+        $try = 1;
+        do {
+            try {
+                $title_append = strtr(__(' (copy {{count}})'), array(
+                    '{{count}}' => $try,
+                ));
+                $clone->form_virtual_name = null;
+                $clone->form_name = $form->title_item().$title_append;
+                if ($clone->save()) {
```
Comment (Meroje) [inline classes/controller/admin/form.ctrl.php]: is this going to overwrite the previous form as you didn't reset the primary key ?

File: lang/fr/common.lang.php [hunk 3]
```diff
@@ -566,4 +566,9 @@
     'A notification of the form answer will be sent to the selected email.' => 'Une notification de la réponse au formulaire sera envoyée à l\'adresse mail sélectionnée.',
     'Label (displayed only in backoffice) :' => 'Label (affiché uniquement sur le back office)',
     '{{form}}: New answer' => '{{form}}&nbsp;: Nouvelle réponse',
+
+    'Slow down, slow down. You have duplicated this form 5 times already. Edit them first before creating more duplicates.' => 'Mollo&nbsp;! Vous avez dupliqué ce formulaire cinq fois déjà. Modifiez les copies existantes avant d’en créer de nouvelles.',
+    'Here you are! The form has just been duplicated.' => 'C’est parti&nbsp;! Le formulaire a été dupliqué.',
+    'The form was not found.' => 'Le formulaire n\'a pas été trouvé.',
```
Comment (Meroje) [inline lang/fr/common.lang.php]: you seem to be one `’` short, have this one

**Review by jguyomard — APPROVED** [review]

**Review by Meroje — COMMENTED** [review]
All these statics don't really need to be on the controller, and I can see the need to duplicate a form outside of the admin appdesk, can you refactor these out on a service class or something ?

File: classes/controller/admin/form.ctrl.php (hunk 2 above)
Comment (tonyyb) [reply classes/controller/admin/form.ctrl.php]: Clone set the item _is_new property to true. 
So saving it make an insert.

File: classes/controller/admin/form.ctrl.php [hunk 4]
```diff
@@ -498,121 +502,59 @@ public function action_duplicate($id = null)
              * @var $form Model_Form
              */
             $form = $this->crud_item($id);
-            if (empty($form->id)) {
-                throw new \Exception(__('The form was not found.'));
+            $contexts = Permission::contexts();
+            $duplicateContext = Input::post('duplicate_context');
```
Comment (Meroje) [inline classes/controller/admin/form.ctrl.php]: Use \Input here

File: views/admin/popup_duplicate.view.php [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        <br/>
+    <?php endforeach; ?>
+    <p style="margin: 1em 0;">
+        <button type="submit" class="ui-priority-primary ui-state-default">
+            <?= __('Duplicate') ?>
+        </button>
+        <span>
+            <?= __('or') ?>
+        </span>
+        <a href="#">
```
Comment (Meroje) [inline views/admin/popup_duplicate.view.php:30]: Should be a button

Comment (tonyyb) [reply views/admin/popup_duplicate.view.php:30]: Why ?

Comment (Meroje) [reply views/admin/popup_duplicate.view.php:30]: It saves you from the effort of cancelling the navigation event, and is a good accessibility habit

Comment (shaoshiva) [reply views/admin/popup_duplicate.view.php:30]: In Novius OS the *Cancel* button should always be a link, that's a choice that was made by the UX team.

And it's not necessarily a bad choice : http://www.lukew.com/ff/entry.asp?571

However it's not the right place to debate this.

**Pull request merged**